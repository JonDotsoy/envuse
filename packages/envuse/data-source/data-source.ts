import { Base } from "./statements/components/base";
import { Block, BlockType } from "./statements/components/block";
import { ArrCursor } from "./statements/lib/arr-cursor";
import { CommentOperator } from "./statements/components/comment-operator";
import { UnexpectedTokenError } from "./statements/tdo/unexpected-token-error";
import { Variable } from "./statements/components/variable";
import { StatementObject } from "./statements/components/statement-object";
import fs from "fs";
import { BlockComment } from "./statements/components/block-comment";
import { CommentInline } from "./statements/components/comment-inline";
import { stringify as DataSourceStringify } from "./statements/lib/stringify";
import { StringifyOptions } from "./statements/lib/stringify-options";

// data source

export type CustomType = {
  type: string;
  parser: (ctx: {
    valueStr: string;
    elementVariable: Variable;
    elementDescription: null | BlockComment;
  }) => any;
};

const stringType: CustomType = {
  type: "string",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    return valueStr;
  },
};

const numberType: CustomType = {
  type: "number",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    const v = Number(valueStr);
    if (isNaN(v) || v === Infinity) {
      ctx.elementVariable.rejectUnexpectedTokenError({
        message: `${valueStr} is not a valid number`,
        position: ctx.elementVariable.valueVariable.pos,
      });
    }
    return v;
  },
};

const booleanType: CustomType = {
  type: "boolean",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    if (["true", "false"].indexOf(valueStr) === -1) {
      ctx.elementVariable.rejectUnexpectedTokenError({
        message: `${valueStr} is not a valid boolean`,
        position: ctx.elementVariable.valueVariable.pos,
      });
    }
    return valueStr === "true";
  },
};

export const defaultCustomTypes: CustomType[] = [
  stringType,
  numberType,
  booleanType,
];

export type Definition = {
  type: string;
  description: string | null;
  value: any;
  valueStr: string;
  elementVariable: Variable;
  elementDescription?: BlockComment | null;
};

export type Option =
  | Buffer
  | {
      filename?: string | null;
      body: Buffer;
    };

export type Values = {
  [k: string]: any;
};

export type CompileOptions = {
  customTypes?: CustomType[];
};

/** AST Parser */
export class DataSource {
  private constructor(private filename: string | null, private body: Buffer) {}

  toAstBody() {
    try {
      return Base.createElement(new Block(this.filename, this.body, 0));
    } catch (err) {
      if (UnexpectedTokenError.isUnexpectedTokenError(err)) {
        // Error.captureStackTrace(err, EnvuseFileParser.prototype.toAstBody);
        // const stack = err.stack;
        // if (stack) {
        //   // let posEndMessage = err.name.length + err.message.length + 3;
        // }
      }
      throw err;
    }
  }

  private static makeOperatorEvaluator(operator: CommentOperator) {
    return (value: { [k: string]: any }) => {
      const a = new ArrCursor(operator.statement.statements);
      let accumulator: any;

      const getValue = (statementObject: StatementObject) => {
        // console.log(inspect(statementObject));
        switch (statementObject.type) {
          case "NameInstance":
            return (
              Array.isArray(statementObject.value) &&
              statementObject.value.reduce(
                (r, path) => (typeof r === "object" ? r[path] : undefined),
                value
              )
            );
          case "Number":
          case "Boolean":
          case "String":
            return statementObject.value;
          default:
            throw new Error("unprocessable statement");
        }
      };

      while (a.has()) {
        if (a.position === 0) {
          accumulator = getValue(a.current());
          a.forward();
          continue;
        }

        switch (a.current().type) {
          case "StrictEqualitySymbol": {
            a.forward();
            // console.log({ accumulator, currentType: a.current().type, currentValue: a.current().value, parsed: getValue(a.current()), value })
            accumulator = accumulator === getValue(a.current());
            a.forward();
            break;
          }
          default:
            throw new Error(`Statement no expected`);
        }
      }

      return accumulator;
    };
  }

  static parse(options: Option & CompileOptions, values?: Values) {
    const ast = this.createDataSource(options);

    const operatorsList = ast.elementList
      .filter(
        (element): element is CommentOperator =>
          element instanceof CommentOperator
      )
      .map((operator) => ({
        operator,
        assert: this.makeOperatorEvaluator(operator),
      }));

    const isVariable = (element: Base): element is Variable =>
      element instanceof Variable;
    const isDescriptor = (element: Base): element is BlockComment =>
      element instanceof BlockComment;
    const isComment = (element: Base): element is CommentInline =>
      element instanceof CommentInline;

    // const variableList = ast.elementList.filter(
    //   (element): element is Variable => element instanceof Variable
    // );

    let description: BlockComment | null = null;

    const variableList = ast.elementList.reduce((acum, element) => {
      if (isDescriptor(element)) {
        description = element;
      }
      if (isComment(element)) {
        description = null;
      }
      if (isVariable(element)) {
        acum.push([element, description]);
        description = null;
      }
      return acum;
    }, [] as [Variable, BlockComment | null][]);

    const definitions = variableList.reduce((acum, [element, description]) => {
      const operators = operatorsList.filter((operator) =>
        operator.operator.elementList.includes(element)
      );

      if (operators.length) {
        const result = operators.reduce(
          (v, operator) =>
            v &&
            operator.assert({
              ...values,
              ...Object.fromEntries(
                Object.entries(acum).map(([k, v]) => [k, v.value])
              ),
            }),
          true
        );
        if (!result) {
          return acum;
        }
      }

      const type = element.typeVariable?.value ?? "string";
      const valueStr = element.valueVariable.value;

      const customTypes = [
        ...defaultCustomTypes,
        ...(options.customTypes ?? []),
      ];

      const getValue = () => {
        for (const customType of customTypes) {
          if (customType.type === type) {
            return customType.parser({
              elementDescription: description,
              elementVariable: element,
              valueStr,
            });
          }
        }

        throw new Error(`Unsupported type ${type}`);
      };

      return {
        ...acum,
        [element.keyVariable.value]: {
          type,
          valueStr,
          description: description?.value ?? null,
          value: getValue(),
          elementVariable: element,
          elementDescription: description,
        },
      };
    }, {} as { [k: string]: Definition });

    const parsed = Object.entries(definitions).reduce(
      (acum, [name, def]) => ({ ...acum, [name]: def.valueStr }),
      {} as { [k: string]: string }
    );

    return { definitions, parsed, ast } as const;
  }

  static parseFile(
    filename: string,
    values?: { [k: string]: any },
    options?: CompileOptions
  ) {
    // read file and store buffer
    const buffer = fs.readFileSync(filename);

    return this.parse({ filename, body: buffer, ...options }, values);
  }

  static createDataSource(options: Option) {
    if (options instanceof Buffer) {
      return new DataSource(null, options).toAstBody();
    }
    return new DataSource(options.filename ?? null, options.body).toAstBody();
  }

  static stringify(comp: BlockType, options?: StringifyOptions) {
    return DataSourceStringify(comp, options);
  }
}
