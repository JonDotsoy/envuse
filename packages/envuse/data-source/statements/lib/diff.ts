import { Base } from "../components/base";
import { Block } from "../components/block";
import { CommentOperator } from "../components/comment-operator";
import { Variable } from "../components/variable";
import util from "util";

class TableOfDiff {
  table = new Map<symbol, (Base | undefined)[]>();

  append(colNum: number, key: symbol, element: Base) {
    const dt = this.table.get(key) ?? [];

    dt[colNum] = element;

    this.table.set(key, dt);
  }
}

const keySymbol = Symbol();

type MapSymbols = {
  [keySymbol]?: symbol;
  [k: string]: MapSymbols | undefined;
};

class KeySymbolMemory {
  private memorySymbols: MapSymbols = {};

  arrToSymbol(k: (string | number)[]) {
    let selected = this.memorySymbols;

    k.forEach((k) => {
      const keyChildren = k;

      const tmpSelected = selected[keyChildren];

      if (tmpSelected) {
        selected = tmpSelected;
      } else {
        const newChildren = {};
        selected[keyChildren] = newChildren;
        selected = newChildren;
      }
    });

    const symbolFound = selected[keySymbol];

    if (symbolFound) return symbolFound;

    const newSymbol = Symbol(util.inspect(k));
    selected[keySymbol] = newSymbol;
    return newSymbol;
  }
}

class KeyGenerator {
  map = new Map<symbol, number>();

  constructor(private keySymbolMemory: KeySymbolMemory) {}

  match(element: Base) {
    const baseKey = [element.$type, elementToKey(element)];
    const keyElement = this.keySymbolMemory.arrToSymbol(baseKey);
    let counter = this.map.get(keyElement) ?? 0;
    this.map.set(keyElement, counter + 1);

    const keyResult = this.keySymbolMemory.arrToSymbol([...baseKey, counter]);

    return keyResult;
  }
}

type Change =
  | {
      type: "deleted";
      base: Base;
    }
  | {
      type: "append";
      compare: Base;
    }
  | {
      type: "changed";
      base: Base;
      compare: Base;
    };

const filterCommentVariable = (
  element: Base
): element is CommentOperator | Variable =>
  element.$type === "CommentOperator" || element.$type === "Variable";

const elementToKey = (element: Base) => {
  if (element instanceof Variable) {
    return element.keyVariable.value;
  } else {
    return element.$type;
  }
};

export const diff = (blockBase: Block, blockCompare: Block) => {
  const keySymbolMemory = new KeySymbolMemory();
  const keyGeneratorBase = new KeyGenerator(keySymbolMemory);
  const keyGeneratorCompare = new KeyGenerator(keySymbolMemory);
  const tableOfDiff = new TableOfDiff();

  const changes: Change[] = [];

  const elementsBase = blockBase.elementList.filter(filterCommentVariable);
  const elementsCompare = blockCompare.elementList.filter(
    filterCommentVariable
  );

  elementsBase.map((element, index) => {
    tableOfDiff.append(0, keyGeneratorBase.match(element), element);
  });

  elementsCompare.map((element, index) => {
    tableOfDiff.append(1, keyGeneratorCompare.match(element), element);
  });

  for (const [elementBase, elementCompare] of tableOfDiff.table.values()) {
    if (elementBase && !elementCompare) {
      changes.push({
        type: "deleted",
        base: elementBase,
      });
      continue;
    } else if (!elementBase && elementCompare) {
      changes.push({
        type: "append",
        compare: elementCompare,
      });
      continue;
    } else if (
      elementBase instanceof Variable &&
      elementCompare instanceof Variable &&
      elementBase.valueVariable.value !== elementCompare.valueVariable.value
    ) {
      changes.push({
        type: "changed",
        base: elementBase,
        compare: elementCompare,
      });
      continue;
    }
  }

  return { changes } as const;
};
