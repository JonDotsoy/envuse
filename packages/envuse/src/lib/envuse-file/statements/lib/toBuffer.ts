export function b(data: string | number[]) {
  if (typeof data === "string") return Buffer.from(data);
  return Buffer.from(data);
}

b.c = Buffer.concat;
