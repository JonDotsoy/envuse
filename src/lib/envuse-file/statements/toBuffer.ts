export function toBuffer(data: string | number[]) {
  if (typeof data === 'string') return Buffer.from(data);
  return Buffer.from(data);
}
