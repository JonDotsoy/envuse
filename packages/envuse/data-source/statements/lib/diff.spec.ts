import { DataSource } from "../../data-source";
import { diff } from "./diff";
import { b } from "./toBuffer";

describe("diff", () => {
  it("should diff deleted", () => {
    const block1 = DataSource.createDataSource(
      Buffer.concat([
        b("foo=bar\n"),
        b("aaa=bbb\n"),
        b("ccc=ddd\n"),
        b("hhh=iii\n"),
      ])
    );

    const block2 = DataSource.createDataSource(
      Buffer.concat([
        b("foo=bar\n"),
        b("aaa=bbb\n"),
        b("eee=fff\n"),
        b("hhh=jjj\n"),
      ])
    );

    expect(diff(block1, block2)).toMatchInlineSnapshot(`
      Object {
        "changes": Array [
          Object {
            "base": Object {
              "$type": "Variable",
              "children": undefined,
              "end": 23,
              "keyVariable": Object {
                "$type": "VariableKey",
                "children": undefined,
                "end": 19,
                "pos": 16,
                "value": "ccc",
              },
              "pos": 16,
              "typeVariable": undefined,
              "valueVariable": Object {
                "$type": "VariableValue",
                "children": undefined,
                "end": 23,
                "pos": 20,
                "value": "ddd",
              },
            },
            "type": "deleted",
          },
          Object {
            "base": Object {
              "$type": "Variable",
              "children": undefined,
              "end": 31,
              "keyVariable": Object {
                "$type": "VariableKey",
                "children": undefined,
                "end": 27,
                "pos": 24,
                "value": "hhh",
              },
              "pos": 24,
              "typeVariable": undefined,
              "valueVariable": Object {
                "$type": "VariableValue",
                "children": undefined,
                "end": 31,
                "pos": 28,
                "value": "iii",
              },
            },
            "compare": Object {
              "$type": "Variable",
              "children": undefined,
              "end": 31,
              "keyVariable": Object {
                "$type": "VariableKey",
                "children": undefined,
                "end": 27,
                "pos": 24,
                "value": "hhh",
              },
              "pos": 24,
              "typeVariable": undefined,
              "valueVariable": Object {
                "$type": "VariableValue",
                "children": undefined,
                "end": 31,
                "pos": 28,
                "value": "jjj",
              },
            },
            "type": "changed",
          },
          Object {
            "compare": Object {
              "$type": "Variable",
              "children": undefined,
              "end": 23,
              "keyVariable": Object {
                "$type": "VariableKey",
                "children": undefined,
                "end": 19,
                "pos": 16,
                "value": "eee",
              },
              "pos": 16,
              "typeVariable": undefined,
              "valueVariable": Object {
                "$type": "VariableValue",
                "children": undefined,
                "end": 23,
                "pos": 20,
                "value": "fff",
              },
            },
            "type": "append",
          },
        ],
      }
    `);
  });
});
