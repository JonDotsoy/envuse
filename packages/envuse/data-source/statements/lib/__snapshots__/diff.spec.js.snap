// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff should diff deleted 1`] = `
Object {
  "changes": Array [
    Object {
      "base": Object {
        "$type": "Variable",
        "children": undefined,
        "end": 24,
        "keyVariable": Object {
          "$type": "VariableKey",
          "children": undefined,
          "end": 19,
          "pos": 16,
          "value": "ccc",
        },
        "pos": 16,
        "typeVariable": undefined,
        "valueVariable": Object {
          "$type": "VariableValue",
          "children": undefined,
          "end": 23,
          "pos": 20,
          "value": "ddd",
        },
      },
      "type": "deleted",
    },
    Object {
      "base": Object {
        "$type": "Variable",
        "children": undefined,
        "end": 32,
        "keyVariable": Object {
          "$type": "VariableKey",
          "children": undefined,
          "end": 27,
          "pos": 24,
          "value": "hhh",
        },
        "pos": 24,
        "typeVariable": undefined,
        "valueVariable": Object {
          "$type": "VariableValue",
          "children": undefined,
          "end": 31,
          "pos": 28,
          "value": "iii",
        },
      },
      "compare": Object {
        "$type": "Variable",
        "children": undefined,
        "end": 32,
        "keyVariable": Object {
          "$type": "VariableKey",
          "children": undefined,
          "end": 27,
          "pos": 24,
          "value": "hhh",
        },
        "pos": 24,
        "typeVariable": undefined,
        "valueVariable": Object {
          "$type": "VariableValue",
          "children": undefined,
          "end": 31,
          "pos": 28,
          "value": "jjj",
        },
      },
      "type": "changed",
    },
    Object {
      "compare": Object {
        "$type": "Variable",
        "children": undefined,
        "end": 24,
        "keyVariable": Object {
          "$type": "VariableKey",
          "children": undefined,
          "end": 19,
          "pos": 16,
          "value": "eee",
        },
        "pos": 16,
        "typeVariable": undefined,
        "valueVariable": Object {
          "$type": "VariableValue",
          "children": undefined,
          "end": 23,
          "pos": 20,
          "value": "fff",
        },
      },
      "type": "append",
    },
  ],
}
`;
